### ISSUE TITLE
Dates and times are not localized and thus displayed only in English

### ISSUE THREAD

**Issue by bastantoine** [LycheeOrg/Lychee#829]
### Detailed description of the problem

When you configure a language other than English, the dates and times are not localized.
This means that for example for a picture taken on 26-07-2020 at 1:25pm, the time and date would be displayed like this _26 July 2020 at 13:25_, even when if you set the language to French (where it should be something like _26 juillet 2020 à 13h25_) or German (where it should be something like _26. Juli 2020 um 13:25 Uhr_).

After a bit of search, I realized that the dates are already formatted on the response of the API. See below a sample of a response of `Album::get` (not related parts have been cut). Same thing goes for `Photo::get`.

```
{
    ....
    "sysdate": "December 2020",
    "min_takestamp": "Jul 2020",
    "max_takestamp": "Jul 2020",
    ...
    "albums": [
        {
            ...
            "sysdate": "December 2020",
            "min_takestamp": "Jul 2020",
            "max_takestamp": "Jul 2020",
           ...
        }
    ],
    ...
}
```

And while browsing through the source code, I found that the dates are formatted using the PHP builtin [`format`](https://www.php.net/manual/fr/datetime.format.php) method, which doesn't seem to support any language other than English (unless you configure your server to the right language, but not this would work when changing it).

From [`app\Models\Album::str_min_takestamp`](https://github.com/LycheeOrg/Lychee/blob/27e4428786c2cacface06c8986d62db02ee0dfaf/app/Models/Album.php#L177):
```php
public function str_min_takestamp()
{
    return $this->min_takestamp == null ? '' : $this->min_takestamp->format('M Y');
}
```

Same thing for for [`app\Models\Album::str_max_takestamp`](https://github.com/LycheeOrg/Lychee/blob/27e4428786c2cacface06c8986d62db02ee0dfaf/app/Models/Album.php#L187) and [`app\Models\Photo::prepare_location_data`](https://github.com/LycheeOrg/Lychee/blob/27e4428786c2cacface06c8986d62db02ee0dfaf/app/Models/Photo.php#L254).

I saw that Carbon provides helpers for the [localization of dates and times](https://carbon.nesbot.com/docs/#api-localization), something like this:
```php
$date = Carbon::now()->locale('fr_FR');

echo $date->locale();            // fr_FR
echo "\n";
echo $date->diffForHumans();     // il y a 1 seconde
echo "\n";
echo $date->monthName;           // décembre
echo "\n";
echo $date->isoFormat('LLLL');   // samedi 19 décembre 2020 23:08
```

Or why not pass only a timestamp and let the front do the localization? It looks like JS already has a method [`toLocaleDateString`](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/Date/toLocaleDateString) on the `Date` class to do that.

**Screenshots**
Note the usage of the French for most of the display, except for the dates.
<img width="294" alt="image" src="https://user-images.githubusercontent.com/8548006/102997212-d59e6980-4524-11eb-9677-03eb9faf2adb.png">

### Output of the diagnostics
    Diagnostics
    -------
    Info: Latest version of PHP is 7.4
    Warning: You may experience problems when uploading a photo of large size. Take a look in the FAQ for details.
    Warning: Dropbox import not working. dropbox_key is empty.

    System Information
    --------------
    Lychee Version (git):            No git data found.
    DB Version:                      4.0.8
    
    composer install:                --no-dev
    APP_ENV:                         production
    APP_DEBUG:                       false
    
    System:                          Linux
    PHP Version:                     7.3
    Max uploaded file size:          20M
    Max post size:                   100M
    MySQL Version:                   10.5.8-MariaDB-1:10.5.8+maria~focal
    
    Imagick:                         1
    Imagick Active:                  1
    Imagick Version:                 1690
    GD Version:                      2.2.5

    Config Information
    --------------
    version:                         040008
    check_for_updates:               0
    sorting_Photos_col:              takestamp
    sorting_Photos_order:            ASC
    sorting_Albums_col:              title
    sorting_Albums_order:            ASC
    imagick:                         1
    skip_duplicates:                 0
    small_max_width:                 0
    small_max_height:                360
    medium_max_width:                1920
    medium_max_height:               1080
    lang:                            fr
    layout:                          1
    image_overlay:                   1
    image_overlay_type:              desc
    default_license:                 none
    compression_quality:             90
    full_photo:                      1
    delete_imported:                 0
    Mod_Frame:                       1
    Mod_Frame_refresh:               30
    thumb_2x:                        1
    small_2x:                        1
    medium_2x:                       1
    landing_page_enable:             0
    landing_owner:                   Bastien ANTOINE
    landing_title:                   Bastien ANTOINE
    landing_subtitle:                Cats, Dogs & Humans Photography
    landing_facebook:                https://www.facebook.com/JohnSmith
    landing_flickr:                  https://www.flickr.com/JohnSmith
    landing_twitter:                 https://www.twitter.com/JohnSmith
    landing_instagram:               https://instagram.com/JohnSmith
    landing_youtube:                 https://www.youtube.com/JohnSmith
    landing_background:              dist/cat.jpg
    site_title:                      Bastien ANTOINE
    site_copyright_enable:           1
    site_copyright_begin:            2017
    site_copyright_end:              2020
    additional_footer_text:          
    display_social_in_gallery:       0
    public_search:                   0
    SL_enable:                       0
    SL_for_admin:                    0
    public_recent:                   0
    recent_age:                      1
    public_starred:                  0
    downloadable:                    0
    photos_wraparound:               1
    map_display:                     0
    zip64:                           1
    map_display_public:              0
    map_provider:                    Wikimedia
    force_32bit_ids:                 0
    map_include_subalbums:           0
    update_check_every_days:         3
    has_exiftool:                    1
    share_button_visible:            0
    import_via_symlink:              0
    has_ffmpeg:                      1
    location_decoding:               0
    location_decoding_timeout:       30
    location_show:                   1
    location_show_public:            0
    rss_enable:                      0
    rss_recent_days:                 7
    rss_max_items:                   100
    prefer_available_xmp_metadata:   0
    editor_enabled:                  1
    lossless_optimization:           0
    swipe_tolerance_x:               150
    swipe_tolerance_y:               250
    local_takestamp_video_formats:   .avi|.mov
    log_max_num_line:                1000
    unlock_password_photos_with_url_param: 0

### Browser and system

macOS Big Sur 11.0.1
Tested on
- Safari 14.0.1
- Chrome 87.0.4280.88

**Comment by kamil4** [comment]
Good analysis! This is something I'm planning to address as part of adding support for editing the take date of the pictures (if I ever actually get around to doing it...).

This is just one instance of a larger issue of Lychee not being fully localized. At least the timestamps are stored in the database in the right format so they can easily be adjusted for display, but other data such as the aperture, focal length, and even the file size are all stored as already formatted strings (`f/2.0`, `28 mm`, `4.3 MB`, etc.). Fixing it would require significant changes to the database and likely the rescan of all the pictures...

**Comment by ildyria** [comment]
related: https://stackoverflow.com/questions/32549845/laravel-5-carbon-global-locale

**Comment by bastantoine** [comment]
> other data such as the aperture, focal length, and even the file size are all stored as already formatted strings

I suppose it helps with the performance to avoid having to refetch them everytime you want to see a picture, but I don't see the problem of localization with them. An aperture of `f/2.0` would be displayed the same in English or in French right? Why do you want to fix them? Plus it's not likely they are going to change, at least the EXIFs.

**Comment by ildyria** [comment]
Aperture, shutter speed, ISO, etc do not need to be translated as they are indeed universal (at least they are the same in French, English, Dutch, German, and Japanese to the extend of my knowledge).

**Comment by kamil4** [comment]
What I meant is that they should've been stored in the database in a "raw" format, i.e., as numbers, not as pre-formatted strings. I agree that there's no need to refetch them every time, but if we decided to fix it, we will probably need to refetch them at least once.

@ildyria, I don't know where your information about the formatting being universal comes from, but a quick browse through a few camera manuals in Russian or, in fact, in your native French, shows that not to be the case. For instance, for shutter speeds in Russian they seem to write "сек" rather than "s". For file sizes it's "МБ" rather than "MB" (and apparently "Mo" in French?). Also, the decimal separator in many non-English countries is a comma, not a dot.

**Comment by bastantoine** [comment]
> shutter speeds in Russian they seem to write "сек" rather than "s"

Looks like this is an issue of localization of the units. After a bit of searching I came across the [Unicode CLDR (Common Locale Data Repository)](http://cldr.unicode.org/index) that seems to be designed to do exactly this (providing translation for some common words/phrases). For example the translation of `MB` in French is given [here](https://github.com/unicode-org/cldr/blob/1cfb084aa6506912f50d47e754232fa0a840978e/common/main/fr.xml#L12159). Same thing for the [Russian](https://github.com/unicode-org/cldr/blob/1cfb084aa6506912f50d47e754232fa0a840978e/common/main/ru.xml#L12383). I guess this will help a lot.

I've found a few PHP libraries that provide APIs to use the CLDR, like [this one](https://punic.github.io/), or [this one](https://github.com/ICanBoogie/CLDR). I don't know which one is better to use though, I haven't tried any of them (yet?).

> the decimal separator in many non-English countries is a comma, not a dot

PHP provides the builtin function [`number_format`](https://www.php.net/manual/en/function.number-format.php) to handle this case. I guess the CLDR provides which decimal and thousand separator to use, or PHP builtins [`localeconv`](https://www.php.net/manual/fr/function.localeconv.php) has it.

Merry Christmas 🥳 !

**Comment by nagmat84** [comment]
> Aperture, shutter speed, ISO, etc do not need to be translated as they are indeed universal (at least they are the same in French, English, Dutch, German, and Japanese to the extend of my knowledge).

I believe that's true. IMHO, to fix the timestamps is the most crucial part, i.e. have some "raw" format in the database (like seconds since epoch or another appropriate encoding) and let the GUI level do the localization.

However, there might be a also some benefit for other attributes, too. For example lets take the shutter speed. Using a fixed base unit only storing a raw multiple of that fix base unit would allow improved sorting and search features. In the long run, one could implement things like "give me all photos which have been taken with at most 2ms shutter speed". At the moment this would not be possible, because the same shutter speed can be stored as "1/500 s", "2ms", "0.002s" and so on. I think the fastest shutter speed commercially available is around 1/32000s. So using microseconds (µs) as a fixed base unit and only store the integer multiple of that base unit in the database would enable such features as mentioned before. However, this is only a "nice to have" and I am not sure, if anybody requires such kind of feature.

Again, the timestamps are the real problem.

**Comment by kamil4** [comment]
Sorry for the delay in responding to this one. As far as I know, the timestamps _in the database_ are not a problem. They are already stored in a raw "timestamp" format. However, the Lychee server currently converts them to English strings before sending them to the frontend, which is indeed a problem that will need fixing.

As to the other fields, as I pointed out in comment https://github.com/LycheeOrg/Lychee/issues/829#issuecomment-750922025, the assertion that the English formatting is universal is not true. But I do agree of course that timestamps are the most critical ones to address, so the rest can wait...

**Comment by nagmat84** [comment]
I assume the issue is solved with [PR #1016](https://github.com/LycheeOrg/Lychee/pull/1016).

Please note, that date/times are now always localized according to the user's browser settings  i.e. in other words they do not respect the explicit language settings of the configuration. However, I assume that this is a minor problem, because I assume that most user will also use a language setting which matches the general local of their OS/browser.
